Add support for named argument functions

eyalleshem opened the pull request:
Add support for named argument function

This commit support functions with argument names.
the format is :
"Select some_function( a => exp, b => exp2 .. ) FROM table1
OR
"select * from table(function(a => exp)) f;"

see:
https://jakewheat.github.io/sql-overview/sql-2016-foundation-grammar.html#named-argument-assignment-token
or example is snwflake db :
https://docs.snowflake.com/en/sql-reference/functions/flatten.html
 This commit does not belong to any branch on this repository.

eyalleshem commented:
Sorry about some style changes that made by `cargo fmt`

nickolay commented:
I find it weird that you added NamedArgument as a new variant of Expr. This makes the parser accept many weird constructs, which shouldn't be accepted. I expected this to be a new SelectItem-like enum. You can use `peek_nth_token(1) == RArrow` to determine if it's a named argument.

The argument name is a simple identifier, not an ObjectName.

Please check the PR for typos (FUCNTION, rigth, agument, // vs /// (which led to rustfmt reformatting the whole enum))

> the format is : "Select * from ( a => exp, b=> exp2 .. )

Surely you meant `SELECT myfunc(args)` rather than `FROM (args)`

eyalleshem commented:
Sorry for all the typos , I fixed them (Hope that all of them..) 
I also changed the the object name to ident - and fixed the commit message with the correct query . 

But i don't sure what you means when you say that the named-argument should be `SelectItem` 
It's a comma-separated argument list that passed to function, In the `struct Function` the args is comma-separated `Vec<Expr>` , And because that my thought was that creating new `Expr` type will be the best suit .

nickolay commented:
I said a _new_ enum, a little like SelectItem (currently with two variants: unnamed argument, named argument).

eyalleshem commented:
I done the changes according to comments .

nickolay commented on src/ast/mod.rs:
```diff
@@ -873,12 +873,42 @@ impl fmt::Display for Assignment {
     }
 }
 
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+pub struct FunctionNamedArg {
+    pub name: Ident,
+    pub arg: Expr,
+}
+
+impl fmt::Display for FunctionNamedArg {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{} => {}", self.name, self.arg)?;
+        Ok(())
+    }
+}
```
Is there a reason to make this a separate struct, rather than `FunctionArg::Named { name, arg }`?

_(Another option is to make the argument type a simple struct with an optional `name`, and the reason I suggested `enum` initially because of the [other possible kinds of SQL arguments](https://jakewheat.github.io/sql-overview/sql-2016-foundation-grammar.html#SQL-argument)...)_

eyalleshem replied:
Fixed , 
I agree that enum is better.

nickolay commented on src/ast/mod.rs:
```diff
@@ -873,12 +873,42 @@ impl fmt::Display for Assignment {
+impl fmt::Display for FunctionArg {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            FunctionArg::Named(named_arg) => named_arg.fmt(f),
+            FunctionArg::Unnamed(unnamed_arg) => unnamed_arg.fmt(f),
```
We use `write!(f, "{}", ...)` in other places for this.

eyalleshem replied:
fixed

nickolay commented on src/parser.rs:
```diff
@@ -2201,11 +2201,24 @@ impl Parser {
+    fn parse_function_args(&mut self) -> Result<FunctionArg, ParserError> {
+        if self.peek_nth_token(1) == Token::RArrow {
+            let name = self.parse_identifier()?;
+            // the `=>`
+            let _ = self.next_token();
```
Please use `expect_token` instead.

eyalleshem replied:
fixed

nickolay commented on src/tokenizer.rs:
```diff
@@ -323,7 +326,7 @@ impl<'a> Tokenizer<'a> {
                         }
                     }
                 }
-                // identifier or keyword
+                // identifier or Keyword
```
Why this change?

eyalleshem replied:
sorry ,  by mistake ..

nickolay commented on src/ast/mod.rs:
```diff
@@ -141,7 +141,7 @@ impl fmt::Display for ObjectName {
 /// An SQL expression of any type.
 ///
-/// The parser does not distinguish between expressions of different types
+/// The parser does not distinguish between qessions of different types
```
Typo?

eyalleshem replied:
sorry ,  by mistake ..

nickolay commented:
Thanks!
